core/bootstrap.go

```diff
@@ -16,108 +17,187 @@ import (
+// - check if an error occurrs (bootstrapping unsuccessful)
+// - wait before starting services which require the node to be bootstrapped
+//
+// If bootstrapping initially fails, Bootstrap() will try again for a total of
+// three times, before giving up completely. Note that in environments where a
+// node may be initialized offline, as normal operation, BootstrapForever()
+// should be used instead.
+//
+// Note: this function could be much cleaner if we were to relax the constraint
+// that we want to exit **after** we have performed initial bootstrapping (and are
+// thus connected to nodes). The constraint may not be that useful in practice.
+// Consider cases when we initialize the node while disconnected from the internet.
+// We don't want this launch to fail... want to continue launching the node, hoping
+// that bootstrapping will work in the future if we get connected.
+func (nb *nodeBootstrapper) TryToBootstrap(ctx context.Context, peers []peer.PeerInfo) error {
+	n := nb.node
+
+	// TODO what bootstrapping should happen if there is no DHT? i.e. we could
+	// continue connecting to our bootstrap peers, but for what purpose? for now
+	// simply exit without connecting to any of them. When we introduce another
+	// routing system that uses bootstrap peers we can change this.
+	dht, ok := n.Routing.(*dht.IpfsDHT)
+	if !ok {
+		return nil
+	}
+
+	for i := 0; i < 3; i++ {
+		if err := bootstrapRound(ctx, n.PeerHost, dht, n.Peerstore, peers); err != nil {
+			return err
+		}
+	}
+
+	// at this point we have done at least one round of initial bootstrap.
+	// we're ready to kick off dht bootstrapping.
+	dbproc, err := dht.Bootstrap(ctx)
+	if err != nil {
+		return err
+	}
+
+	// kick off the node's periodic bootstrapping
+	proc := periodicproc.Tick(BootstrapPeriod, func(worker goprocess.Process) {
+		if err := bootstrapRound(ctx, n.PeerHost, dht, n.Peerstore, peers); err != nil {
 			log.Error(err)
 		}
-		select {
-		case <-parent.Done():
-			return parent.Err()
-		case <-time.Tick(period):
+	})
+
+	// add dht bootstrap proc as a child, so it is closed automatically when we are.
+	proc.AddChild(dbproc)
+
+	// we were given a context. instead of returning proc for the caller
+	// to manage, for now we just close the proc when context is done.
+	go func() {
+		<-ctx.Done()
+		proc.Close()
+	}()
+	return nil
+}
+
+// BootstrapForever starts IpfsNode bootstrapping. Unlike TryToBootstrap(),
+// BootstrapForever() will run indefinitely (until its context is cancelled).
+// This is particularly useful for the daemon and other services, which may
+// be started offline and will come online at a future date.
+//
+// TODO: check offline --to--> online case works well and doesn't hurt perf.
+// We may still be dialing. We should check network config.
+func (nb *nodeBootstrapper) BootstrapForever(ctx context.Context, peers []peer.PeerInfo) error {
+	for {
+		if err := nb.TryToBootstrap(ctx, peers); err == nil {
```

Comment:
so, in the case where the DHT is there and TryToBootstrap returns nil, this is just gonna loop at full tilt forever?